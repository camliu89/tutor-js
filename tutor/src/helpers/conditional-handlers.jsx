import React from 'react';
import invariant from 'invariant';
import { extend, last, first } from 'lodash';
import { Redirect } from 'react-router-dom';
import { asyncComponent } from './async-component';
import { CourseNotFoundWarning } from '../components/course-not-found-warning';
import Courses from '../models/courses-map';
import { OXMatchByRouter } from 'shared';

const StudentDashboard = asyncComponent(
  () => import('../screens/student-dashboard')
);

const getConditionalHandlers = (Router) => {
  const MatchForTutor = OXMatchByRouter(Router, null, 'TutorRouterMatch');

  const renderDashboard = (props) => {
    const { courseId } = props.params;

    extend(props, { courseId });
    const course = Courses.get(courseId);

    if (!course) {
      return <CourseNotFoundWarning />;
    }

    if (!props.match.isExact) {
      return (
        <MatchForTutor {...props} />
      );
    }

    if (course.isTeacher) {
      if (course.is_concept_coach) {
        return (
          <CCDashboard courseId={courseId} {...props} />
        );
      } else {
        return (
          <Redirect
            to={{
              pathname: Router.makePathname('viewTeacherDashboard', props.params),
              query: Router.currentQuery(),
            }} />
        );
      }
    } else {
      if (course.is_concept_coach) {
        return (
          <CCStudentRedirect courseId={courseId} />
        );
      } else {
        return <StudentDashboard {...props} />;
      }
    }
  };
  // eslint-disable-next-line react/prop-types
  const legacyReferenceBookRedirect = ({ params }) => {
    const parts = params.parts.split('/');
    const course = Courses.get(first(parts));
    invariant(course, `Did not find course for params '${params}'`);
    const chapterSection = parts.length > 1 ? last(parts) : 0;
    return (
      <Redirect
        to={{
          pathname: Router.makePathname(
            chapterSection ? 'viewReferenceBookSection' : 'viewReferenceBook',
            {
              ecosystemId: course.ecosystem_id,
              chapterSection,
            }),
          query: Router.currentQuery(),
        }}
      />
    );
  };

  return {
    dashboard() { return renderDashboard; },
    legacyReferenceBookRedirect() { return legacyReferenceBookRedirect; },
  };
};

export { getConditionalHandlers };
