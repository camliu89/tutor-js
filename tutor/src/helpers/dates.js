import moment from 'moment';
import 'twix';
import 'moment-timezone';
import { compact, reduce, min, max, map } from 'lodash';
import TimeHelper from './time';

export function getDay(oneMoment) {
  return moment(oneMoment)
    .startOf('day')
    .twix(moment(oneMoment).endOf('day'), { allDay: true });
}

export function getDurationFromMoments(listOfMoments) {
  return reduce(
    listOfMoments,
    (current, next) => current.union(getDay(next)),
    getDay(listOfMoments[0])
  );
}

export function findEarliest(dateThings) {
  return min(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}

export function findLatest(dateThings) {
  return max(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}


export function dateWithUnchangedTime(date, t) {
  if (moment.isMoment(date)) {
    date = date.format(TimeHelper.ISO_DATE_FORMAT);
  }
  return TimeHelper.getDateOnly(date);
}
