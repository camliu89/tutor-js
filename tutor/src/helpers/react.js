import React from 'react';
import ReactDOM from 'react-dom';
import cn from 'classnames';
import { propTypes as mobxPropTypes } from 'mobx-react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import styled, { css } from 'styled-components';
import { idType } from 'shared';

export {
  React, cn, ReactDOM, mobxPropTypes, PropTypes,
  invariant, styled, css, idType,
};
export { observable, action, computed } from 'mobx';
export { observer, inject, Provider } from 'mobx-react';
