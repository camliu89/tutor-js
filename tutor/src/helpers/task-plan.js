import { isEmpty, get, pick, first, every, reduce, map } from 'lodash';
import moment from 'moment';
import Time from '../models/time';

export default {

  // returns the opens_at and due_at dates combined if they're identical,
  // separated by target_id if not
  dates(plan, options) {
    if (options == null) { options = {}; }
    if (isEmpty( get(plan, 'tasking_plans' ))) { return {}; }
    const pickFn = options.only ? tp => pick(tp, options.only) : tp => pick(tp, 'opens_at', 'due_at');

    const firstTasking = first( plan.tasking_plans );
    const isIdentical = every(plan.tasking_plans, function(tp) {
      if (options.only != null) {
        return tp[options.only] === firstTasking[options.only];
      } else {
        return (tp.opens_at === firstTasking.opens_at) && (tp.due_at === firstTasking.due_at);
      }
    });
    if (isIdentical) {
      return { all: pickFn(firstTasking) };
    } else {
      return reduce( plan.tasking_plans, function(memo, tp) {
        memo[tp.target_id] = pickFn(tp);
        return memo;
      }
      , {});
    }
  },

  calendarParams(course) {
    let date;
    const dueAtRange = course
      .teacherTaskPlans.array
      .map(plan => plan.dateRanges.due.start)
      .sort();

    if (dueAtRange.length) {
      date = dueAtRange[0];
    } else {
      date = Time.now;
    }
    return {
      to: 'calendarByDate',
      params: {
        courseId: course.id,
        date: moment(date).format('YYYY-MM-DD'),
      },
    };
  },

  earliestDueDate(plan) {
    const dates = map(get(plan, 'tasking_plans'), 'due_at');
    return first(dates.sort()) || '';
  },

};
