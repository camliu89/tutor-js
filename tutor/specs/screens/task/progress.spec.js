import UX from '../../../src/screens/task/ux';
import Progress from '../../../src/screens/task/progress';

import { Factory } from '../../helpers';

describe('Reading Progress Component', () => {
  let props;

  beforeEach(() => {
    const task = Factory.studentTasks({
      count: 1,
      attributes: { type: 'reading', stepCount: 2 },
    }).array[0];
    props = {
      ux: new UX({ task }),
    };
  });

  it('matches snapshot', () => {
    expect(<Progress {...props}><p>hi</p></Progress>).toMatchSnapshot();
  });

  it('pages through steps', () => {
    const pr = mount(<Progress {...props}><p>hi</p></Progress>);
    expect(pr).toHaveRendered('a.paging-control.prev[disabled=true]');
    expect(pr).toHaveRendered('a.paging-control.next[disabled=false]');
    pr.find('a.paging-control.next').simulate('click');
    expect(pr).toHaveRendered('a.paging-control.next[disabled=true]');
    expect(pr).toHaveRendered('a.paging-control.prev[disabled=false]');
    expect(props.ux.stepIndex).toEqual(1);
    pr.unmount();
  });

});
