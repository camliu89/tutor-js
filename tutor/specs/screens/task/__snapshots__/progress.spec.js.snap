// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reading Progress Component matches snapshot 1`] = `
<ProgressCard
  ux={
    TaskUX {
      "_stepIndex": 0,
      "_task": StudentTask {
        "complete": undefined,
        "description": undefined,
        "due_at": 2019-04-23T05:00:00.000Z,
        "feedback_at": undefined,
        "id": 1,
        "steps": Array [
          StudentTaskStep {
            "$nonSerializable": Object {
              "task": true,
            },
            "answer_id": undefined,
            "content": StudentTaskInteractiveStep {
              "chapter_section": Array [
                1,
                2,
              ],
              "content_html": "<span class=\\"os-text\\">Physics: An Introduction</span> </h2> <div class=\\"os-figure\\"> <figure id=\\"import-auto-id1580373\\"><span data-alt=\\"Two Canada geese flying close to each other in the sky.\\" data-type=\\"media\\" id=\\"import-auto-id1936836\\"><img alt=\\"Two Canada geese flying close to each other in the sky.\\" data-media-type=\\"image/wmf\\" id=\\"90580\\" src=\\"https://cnx.org/resources/18f7913be5bf22af357a34143dd8638511b9ba9a\\" width=\\"325\\"></span> </figure><div class=\\"os-caption-container\\"> <span class=\\"os-title-label\\">Figure </span><span class=\\"os-number\\">1.2</span><span class=\\"os-divider\\"s as food calories, batteries, heat, light, and watch springs. Understanding this law makes it easier to learn about the various forms energy takes and how they relate to one another. Apparently unrelated topics are connected through broadly applicable physical laws, permitting an understanding beyond just the memorization of lists of facts.</p> <p id=\\"import-auto-id2555094\\">The unifying aspect of physical laws and the basic simplicity",
              "has_learning_objectives": false,
              "id": 1,
              "related_content": Array [
                Object {
                  "chapter_section": Array [
                    1,
                    2,
                  ],
                  "title": "Food Irradiation",
                },
              ],
              "title": "Electric Hazards and the Human Body",
              "type": "interactive",
            },
            "correct_answer_id": undefined,
            "feedback_html": undefined,
            "formats": Array [],
            "free_response": undefined,
            "id": 1,
            "isFetched": true,
            "is_completed": false,
            "labels": Array [],
            "multiPartGroup": undefined,
            "preview": "Sound Interference and Resonance: Standing Waves in Air Columns",
            "task": [Circular],
            "task_association_name": "steps",
            "type": "interactive",
            "uid": null,
          },
          StudentTaskStep {
            "$nonSerializable": Object {
              "task": true,
            },
            "answer_id": undefined,
            "content": StudentTaskExerciseStep {
              "_page": undefined,
              "average_step_number": undefined,
              "book": undefined,
              "content": Exercise {
                "$nonSerializable": Object {
                  "wrapper": true,
                },
                "attachments": Array [],
                "authors": Array [],
                "copyright_holders": Array [],
                "id": undefined,
                "is_vocab": undefined,
                "nickname": undefined,
                "number": 11945,
                "published_at": undefined,
                "questions": Array [
                  ExerciseQuestion {
                    "$nonSerializable": Object {
                      "exercise": true,
                    },
                    "answers": Array [
                      ExerciseAnswer {
                        "$nonSerializable": Object {
                          "question": true,
                        },
                        "content_html": "geriatrics",
                        "correctness": undefined,
                        "feedback_html": undefined,
                        "id": "259301",
                        "question": [Circular],
                        "question_association_name": "answers",
                      },
                      ExerciseAnswer {
                        "$nonSerializable": Object {
                          "question": true,
                        },
                        "content_html": "gerontology",
                        "correctness": undefined,
                        "feedback_html": undefined,
                        "id": "259302",
                        "question": [Circular],
                        "question_association_name": "answers",
                      },
                      ExerciseAnswer {
                        "$nonSerializable": Object {
                          "question": true,
                        },
                        "content_html": "secondary aging",
                        "correctness": undefined,
                        "feedback_html": undefined,
                        "id": "259303",
                        "question": [Circular],
                        "question_association_name": "answers",
                      },
                      ExerciseAnswer {
                        "$nonSerializable": Object {
                          "question": true,
                        },
                        "content_html": "primary aging",
                        "correctness": undefined,
                        "feedback_html": undefined,
                        "id": "259304",
                        "question": [Circular],
                        "question_association_name": "answers",
                      },
                    ],
                    "collaborator_solutions": Array [],
                    "exercise": [Circular],
                    "exercise_association_name": "questions",
                    "formats": Array [
                      ExerciseFormat {
                        "$nonSerializable": Object {
                          "question": true,
                        },
                        "_format": "free-response",
                        "question": [Circular],
                        "question_association_name": "formats",
                      },
                      ExerciseFormat {
                        "$nonSerializable": Object {
                          "question": true,
                        },
                        "_format": "multiple-choice",
                        "question": [Circular],
                        "question_association_name": "formats",
                      },
                    ],
                    "hints": Array [],
                    "id": "62902",
                    "is_answer_order_important": false,
                    "stem_html": "In the early 1900s, a New York physician named Dr. Ignatz Nascher developed a medical specialty that focused on the elderly. This field is known broadly as ___.",
                    "stimulus_html": "",
                  },
                ],
                "stimulus_html": undefined,
                "tags": Array [
                  ExerciseTag {
                    "_tag": "time:short",
                  },
                  ExerciseTag {
                    "_tag": "dok:2",
                  },
                  ExerciseTag {
                    "_tag": "blooms:2",
                  },
                  ExerciseTag {
                    "_tag": "type:practice",
                  },
                  ExerciseTag {
                    "_tag": "requires-context:true",
                  },
                  ExerciseTag {
                    "_tag": "book:stax-soc",
                  },
                  ExerciseTag {
                    "_tag": "lo:stax-soc:13-2-2",
                  },
                  ExerciseTag {
                    "_tag": "context-cnxmod:d9df5e48-4b72-482e-b616-886926188054",
                  },
                  ExerciseTag {
                    "_tag": "context-cnxfeature:fs-id2332825",
                  },
                ],
                "uid": "11945@4",
                "uuid": "7ddae47d-c4ec-437e-80bd-3f54d8a639a7",
                "version": 4,
                "versions": Array [],
                "wrapper": [Circular],
                "wrapper_association_name": "content",
              },
              "context": "<div data-type=\\"note\\" data-has-label=\\"true\\" id=\\"fs-id2332825\\" class=\\"sociological-research\\" a-type=\\"term\\" id=\\"import-auto-id1169033058271\\">geriatrics</span>, a medical specialty that focuses on the elderly. He created the word by combining were before Nascherâ€™s ideas gained acceptance?</p> </div></div> </div>",
              "course": undefined,
              "ecosystem_id": undefined,
              "has_interactive": false,
              "has_video": false,
              "id": 1,
              "isSelected": false,
              "is_excluded": false,
              "page_uuid": false,
              "pool_types": Array [],
              "preview": undefined,
              "question_stats": Array [],
              "tags": Array [],
              "url": "",
            },
            "correct_answer_id": undefined,
            "feedback_html": undefined,
            "formats": Array [
              "multiple-choice",
            ],
            "free_response": undefined,
            "id": 1,
            "isFetched": true,
            "is_completed": false,
            "labels": Array [],
            "multiPartGroup": undefined,
            "preview": "clicks-and-mortar: Angiosperms",
            "task": [Circular],
            "task_association_name": "steps",
            "type": "exercise",
            "uid": "2@1",
          },
        ],
        "tasksMap": StudentTasks {
          "_map": Object {
            "1": [Circular],
          },
          "course": Course {
            "appearance_code": "college_physics",
            "book_pdf_url": "https://archive.cnx.org/exports/185cbf87-c72e-48f5-b51e-f14f21b5eabd.pdf",
            "cloned_from_id": null,
            "default_due_time": "07:00",
            "default_open_time": "00:32",
            "does_cost": false,
            "ecosystem_book_uuid": "72614e6c-ebf8-4879-9145-76ade4177f38",
            "ecosystem_id": 1,
            "ends_at": "2019-05-20T05:00:00.000Z",
            "homework_progress_weight": 0.1,
            "homework_score_weight": 0.4,
            "id": 1,
            "is_access_switchable": true,
            "is_active": true,
            "is_college": true,
            "is_concept_coach": false,
            "is_lms_enabled": undefined,
            "is_lms_enabling_allowed": false,
            "is_preview": false,
            "just_created": false,
            "map": undefined,
            "name": "College Physics",
            "offering_id": 1,
            "periods": Array [
              CoursePeriod {
                "$nonSerializable": Object {
                  "course": true,
                },
                "course": [Circular],
                "course_association_name": "periods",
                "default_due_time": "07:00",
                "default_open_time": "00:01",
                "enrollment_code": "66861",
                "enrollment_url": "http://localhost:3001/enroll/66861",
                "id": 1,
                "is_archived": false,
                "name": "1st",
                "num_enrolled_students": 38,
                "teacher_student_role_id": 1,
              },
              CoursePeriod {
                "$nonSerializable": Object {
                  "course": true,
                },
                "course": [Circular],
                "course_association_name": "periods",
                "default_due_time": "07:00",
                "default_open_time": "00:01",
                "enrollment_code": "66861",
                "enrollment_url": "http://localhost:3001/enroll/66861",
                "id": 2,
                "is_archived": false,
                "name": "1st",
                "num_enrolled_students": 51,
                "teacher_student_role_id": 2,
              },
              CoursePeriod {
                "$nonSerializable": Object {
                  "course": true,
                },
                "course": [Circular],
                "course_association_name": "periods",
                "default_due_time": "07:00",
                "default_open_time": "00:01",
                "enrollment_code": "66861",
                "enrollment_url": "http://localhost:3001/enroll/66861",
                "id": 3,
                "is_archived": false,
                "name": "1st",
                "num_enrolled_students": 62,
                "teacher_student_role_id": 3,
              },
            ],
            "reading_progress_weight": 0.2,
            "reading_score_weight": 0.2,
            "roles": Array [
              CourseRole {
                "id": 1,
                "joined_at": 2019-03-27T05:00:00.000Z,
                "research_identifier": undefined,
                "type": "student",
              },
            ],
            "salesforce_book_name": "College Physics",
            "starts_at": "2019-03-20T05:00:00.000Z",
            "students": Array [
              CourseStudent {
                "$nonSerializable": Object {
                  "course": true,
                },
                "course": [Circular],
                "course_association_name": "students",
                "first_name": "Blaze",
                "id": 1,
                "is_active": true,
                "is_comped": false,
                "is_paid": false,
                "is_refund_allowed": undefined,
                "is_refund_pending": undefined,
                "last_name": "Kris",
                "name": undefined,
                "payment_due_at": 2019-04-23T05:00:00.000Z,
                "period_id": undefined,
                "prompt_student_to_pay": false,
                "role_id": 1,
                "roster": undefined,
                "student_identifier": "",
                "uuid": undefined,
              },
            ],
            "term": "spring",
            "time_zone": "Central Time (US & Canada)",
            "uuid": "18b86a4d-7bb9-45b6-9a55-334fac1c6062",
            "webview_url": "https://qa.cnx.org/contents/185cbf87-c72e-48f5-b51e-f14f21b5eabd",
            "year": 2017,
          },
          "keyType": [Function],
        },
        "title": "solution-oriented",
        "type": "reading",
      },
      "course": Course {
        "appearance_code": "college_physics",
        "book_pdf_url": "https://archive.cnx.org/exports/185cbf87-c72e-48f5-b51e-f14f21b5eabd.pdf",
        "cloned_from_id": null,
        "default_due_time": "07:00",
        "default_open_time": "00:32",
        "does_cost": false,
        "ecosystem_book_uuid": "72614e6c-ebf8-4879-9145-76ade4177f38",
        "ecosystem_id": 1,
        "ends_at": "2019-05-20T05:00:00.000Z",
        "homework_progress_weight": 0.1,
        "homework_score_weight": 0.4,
        "id": 1,
        "is_access_switchable": true,
        "is_active": true,
        "is_college": true,
        "is_concept_coach": false,
        "is_lms_enabled": undefined,
        "is_lms_enabling_allowed": false,
        "is_preview": false,
        "just_created": false,
        "map": undefined,
        "name": "College Physics",
        "offering_id": 1,
        "periods": Array [
          CoursePeriod {
            "$nonSerializable": Object {
              "course": true,
            },
            "course": [Circular],
            "course_association_name": "periods",
            "default_due_time": "07:00",
            "default_open_time": "00:01",
            "enrollment_code": "66861",
            "enrollment_url": "http://localhost:3001/enroll/66861",
            "id": 1,
            "is_archived": false,
            "name": "1st",
            "num_enrolled_students": 38,
            "teacher_student_role_id": 1,
          },
          CoursePeriod {
            "$nonSerializable": Object {
              "course": true,
            },
            "course": [Circular],
            "course_association_name": "periods",
            "default_due_time": "07:00",
            "default_open_time": "00:01",
            "enrollment_code": "66861",
            "enrollment_url": "http://localhost:3001/enroll/66861",
            "id": 2,
            "is_archived": false,
            "name": "1st",
            "num_enrolled_students": 51,
            "teacher_student_role_id": 2,
          },
          CoursePeriod {
            "$nonSerializable": Object {
              "course": true,
            },
            "course": [Circular],
            "course_association_name": "periods",
            "default_due_time": "07:00",
            "default_open_time": "00:01",
            "enrollment_code": "66861",
            "enrollment_url": "http://localhost:3001/enroll/66861",
            "id": 3,
            "is_archived": false,
            "name": "1st",
            "num_enrolled_students": 62,
            "teacher_student_role_id": 3,
          },
        ],
        "reading_progress_weight": 0.2,
        "reading_score_weight": 0.2,
        "roles": Array [
          CourseRole {
            "id": 1,
            "joined_at": 2019-03-27T05:00:00.000Z,
            "research_identifier": undefined,
            "type": "student",
          },
        ],
        "salesforce_book_name": "College Physics",
        "starts_at": "2019-03-20T05:00:00.000Z",
        "students": Array [
          CourseStudent {
            "$nonSerializable": Object {
              "course": true,
            },
            "course": [Circular],
            "course_association_name": "students",
            "first_name": "Blaze",
            "id": 1,
            "is_active": true,
            "is_comped": false,
            "is_paid": false,
            "is_refund_allowed": undefined,
            "is_refund_pending": undefined,
            "last_name": "Kris",
            "name": undefined,
            "payment_due_at": 2019-04-23T05:00:00.000Z,
            "period_id": undefined,
            "prompt_student_to_pay": false,
            "role_id": 1,
            "roster": undefined,
            "student_identifier": "",
            "uuid": undefined,
          },
        ],
        "term": "spring",
        "time_zone": "Central Time (US & Canada)",
        "uuid": "18b86a4d-7bb9-45b6-9a55-334fac1c6062",
        "webview_url": "https://qa.cnx.org/contents/185cbf87-c72e-48f5-b51e-f14f21b5eabd",
        "year": 2017,
      },
      "router": TestRouter {
        "history": Object {
          "createHref": [MockFunction],
          "push": [MockFunction] {
            "calls": Array [
              Array [
                "/course/1/task/1/step/1",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
          "replace": [MockFunction],
        },
        "route": Object {
          "hash": "",
          "location": Object {
            "pathname": "/",
          },
          "match": Object {
            "params": Object {},
          },
          "search": "",
        },
      },
      "window": [Window],
    }
  }
>
  <p>
    hi
  </p>
</ProgressCard>
`;
